lib/Dialect/FIRRTL/Transforms/AddSeqMemPorts.cpp

```diff
@@ -308,22 +313,43 @@ void AddSeqMemPortsPass::createOutputFile(igraph::ModuleOpInterface module) {
 
   // The current sram we are processing.
   unsigned sramIndex = 0;
-  auto dutSymbol = FlatSymbolRefAttr::get(module.getModuleNameAttr());
   auto &instancePaths = memInfoMap[module].instancePaths;
-  for (auto instancePath : instancePaths) {
-    os << sramIndex++ << " -> ";
-    addSymbol(dutSymbol);
-    // The instance path is reverse from the order we print it.
-    for (auto ref : llvm::reverse(instancePath)) {
-      os << ".";
-      addSymbol(ref);
-    }
-    os << "\n";
-  }
+  auto dutSymbol = FlatSymbolRefAttr::get(module.getModuleNameAttr());
 
-  // Put the information in a verbatim operation.
   auto loc = builder.getUnknownLoc();
+  // Builder to create the hierpath to represent the path to the sram.
+  auto nlaBuilder =
+      mlir::ImplicitLocOpBuilder::atBlockBegin(loc, circuit.getBodyBlock());
+  // Put the information in a verbatim operation.
   builder.create<emit::FileOp>(loc, outputFile, [&] {
+    for (auto instancePath : instancePaths) {
+      // Note: Reverse instancepath to construct the NLA.
+      SmallVector<Attribute> path(llvm::reverse(instancePath));
+      os << sramIndex++ << " -> ";
+      addSymbol(dutSymbol);
+      os << ".";
+
+      auto nla = nlaBuilder.create<hw::HierPathOp>(
+          loc, builder.getStringAttr(circtNamespace.newName("memNLA")),
+          builder.getArrayAttr(path));
+      auto nlaSymbol = FlatSymbolRefAttr::get(nla);
+      addSymbol(nlaSymbol);
+      NamedAttrList fields;
+      // There is no current client for the distinct attr, but it will be used
+      // by OM::path once the metadata is moved to OM, instead of the verbatim.
+      auto id = DistinctAttr::create(UnitAttr::get(builder.getContext()));
+      fields.append("id", id);
+      fields.append("class", builder.getStringAttr("circt.tracker"));
+      fields.append("circt.nonlocal", nlaSymbol);
+      // Now add the nonlocal annotation to the leaf instance.
+      auto *leafInstance = innerRefToInstanceMap[instancePath.front()];
+      // TODO: Check if we can reuse HierPathOp, if it already exists!
+      AnnotationSet annos(leafInstance);
```

Comment:
Checkout `HierPathCache` from `FIRRTLAnnotationHelper.h`! 😁

Reply:
Awesome, thanks I missed `HierPathCache`. Updated the PR to use the cache.